samples/infix/reader.cc

```diff
@@ -96,18 +184,89 @@ namespace
             (T(Group) << (T(Print) * T(String)[Lhs] * Any++[Rhs])) >>
           [](Match& _) { return Output << _(Lhs) << (Expression << _[Rhs]); },
 
+        // --- tuples only:
+
+        // an append operation looks like append(...); we rely on tuple parsing
+        // to get the (...) part right, but we can immediately recognize
+        // append with a () next to it.
+        // Marking the (...) under Expression ensures normal parsing happens to
+        // it later.
+        In(Expression) * T(Append)[Append] * T(Paren)[Paren] >>
+          [](Match& _) {
+            return Expression << (_(Append) << (Expression << _(Paren)));
+          },
+
+        // (...), a group with parens, might be a tuple.
+        // It depends on how many commas there are in it.
+        // We need this if tuples require parens, and we are not handling tuples
+        // in the parser.
+        In(Expression) *
+            (T(Paren)[Paren]
+             << T(Group)[Group])(enable_if_parens_no_parser_tuples) >>
+          [](Match& _) {
+            return Expression << (TupleCandidate ^ _(Paren)) << *_[Group];
+          },
+        // Special case: empty () is rejected without tuples, but we accept it
+        // as an empty tuple.
+        In(Expression) *
+            (T(Paren)[Paren] << End)(enable_if_parens_no_parser_tuples) >>
+          [](Match& _) { return Expression << (TupleCandidate ^ _(Paren)); },
+
+        // --- tuples only, parser tuples version:
+        // Special case: a ParserTuple with one empty group child is (,), which
+        // we consider a valid empty tuple.
+        In(Expression) *
+            (T(Paren)
+             << (T(ParserTuple)[ParserTuple] << ((T(Group) << End) * End))) >>
+          [](Match& _) { return Expression << (ParserTuple ^ _(ParserTuple)); },
+        // Special case 2: a ParserTuple with no children is (), which is also a
+        // valid empty tuple.
+        In(Expression) * (T(Paren)[Paren] << End)(enable_if_parser_tuples) >>
+          [](Match& _) { return Expression << (ParserTuple ^ _(Paren)); },
+        // This grabs any Paren whose only child is a ParserTuple, and unwraps
+        // that ParserTuple's nested groups into Expression nodes so future
+        // passes will parse their contents properly.
+        In(Expression) *
```

Comment:
I would say that the method you suggest is less self-documenting.

The way I've been writing (and this includes some C++ we're also talking about), I prefer code that reads like a list of self-describing cases. Maybe it's a disjunction of possible string patterns, or maybe it's a list of shapes a valid ParserTuple is allowed to have.

Splitting up the pattern here takes a relatively readable for loop and a list of 3 patterns that exhaustively describe valid ParserTuples, and mixes them together so that you can neither see the for loop nor easily read the ParserTuple cases.

Put another way, reading goes from a syntactic enumeration to a flow-sensitive analysis.

Reply:
TODO: write this as comment

Reply:
Done in next commit.